dev/io.openliberty.checkpoint/src/io/openliberty/checkpoint/internal/CheckpointImpl.java

```diff
@@ -110,19 +103,23 @@ public void failedCompletion(Future<Boolean> future, Throwable t) {
 
     @Override
     public void check() {
-        if (checkpointApplications) {
-            try {
-                checkpoint(Phase.APPLICATIONS);
-            } catch (CheckpointFailed e) {
-                if (e.getType() == Type.SNAPSHOT_FAILED) {
-                    new Thread(() -> System.exit(e.getErrorCode()), "Checkpoint exit.").start();
-                }
-                // TODO should we always exit on failure?
-            }
+        if (checkpointAt == Phase.APPLICATIONS) {
+            checkpointOrExitOnFailure(Phase.APPLICATIONS);
         }
     }
 
-    void checkpoint(Phase phase) throws CheckpointFailed {
+    void checkpointOrExitOnFailure(Phase phase) {
+        try {
+            checkpoint(phase);
+        } catch (CheckpointFailedException e) {
+            // TODO log error informing we are exiting
+
+            // TODO is there any type of failure where we would not want to exit?
+            new Thread(() -> System.exit(e.getErrorCode()), "Checkpoint failed, exiting...").start();
```

Comment:
Can you add a comment as to why you are launching a new Thread? I think you said it avoids a quiesce wait?

Reply:
I'll add a comment.  The extra thread is needed to avoid blocking the current thread while the shutdown hooks are run (which starts a server shutdown and quiesce).  The issue is that if we block the event thread then there is a deadlock that prevents the server shutdown from happening until a 30 second timeout is reached.